src/nowsecure-snapshot.ts

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright © 2022 NowSecure Inc.
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+import type { DependencySnapshot, Manifest } from "./types/dependency-snapshot";
+import type { Deputy } from "./types/deputy";
+import * as client from "@actions/http-client";
+import { version } from "./nowsecure-version";
+import { USER_AGENT } from "./nowsecure-client";
+
+/**
+ * Contains environmental information for conversion routines.
+ */
+export interface ActionContext {
+  sha: string;
+  ref: string;
+  job: string;
+  runId: number;
+  repo: {
+    owner: string;
+    repo: string;
+  };
+}
+
+/**
+ * Upload URL for snapshots.
+ */
+const snapshotUrl = (owner: string, repo: string) =>
+  `https://api.github.com/repos/${owner}/${repo}/dependency-graph/snapshots`;
+
+function encodePurl(ecosystem: string, name: string, version?: string): string {
+  let purl = `pkg:${ecosystem}/${name}`;
+  if (version) {
+    purl += `@${version}`;
+  }
+
+  return purl;
+}
+
+/**
+ * Convert a Platform report to the Snapshot format.
+ */
+export function convertToSnapshot(
+  deputy: Deputy,
+  { ref, sha, job, runId }: ActionContext
+): DependencySnapshot {
+  // The file name of the application bundle that was scanned.
+  const name = deputy.package;
+  const manifests = new Map<string, Manifest>();
+
+  for (const component of deputy.components) {
+    // If there is a leading slash, remove.
+    let source = component.source;
+    if (source.length > 0 && source[0] === "/") {
+      source = source.substr(1);
+    }
+
+    if (!manifests.has(source)) {
+      const manifest: Manifest = {
+        name: source,
+        resolved: {},
+      };
+
+      manifests.set(source, manifest);
+    }
+
+    const manifest = manifests.get(source);
+
+    const purl = encodePurl(
+      component.ecosystem.toLowerCase(),
+      component.name,
+      component.version
+    );
+
+    manifest.resolved[component.name] = {
+      purl: purl.toString(),
+      relationship: "direct",
+    };
+
+    manifests.set(source, manifest);
+  }
+
+  const snapshot: DependencySnapshot = {
+    version: 0,
+    ref,
+    sha,
+    job: {
+      name: job,
+      id: `${runId}`,
+    },
+    detector: {
+      name: "NowSecure",
+      url: "https://www.nowsecure.com/",
+      version,
+    },
+    scanned: new Date().toISOString(),
+    manifests: Object.fromEntries(manifests),
+  };
+
+  return snapshot;
+}
+
+/**
+ * Submit a Snapshot to GitHub.
+ */
+export async function submitSnapshotData(
```

Comment:
just noting that the nowsecure-client way of doing this via a class and the way we're doing it here via a function are a bit different. leads me to at least ask the question, should this be a method on the nowsecure client? why not?

(for example we set things like retries on the other client, but looks like we won't here).

Reply:
The intention is that this code is temporary: once GitHub releases an `upload-sarif`-stlye Action for SBOM we will remove this code and instead write it to a file and have the user use the GitHub Action. It doesn't make sense to add this to the NowSecure client class, since that is used for communication with NowSecure services only.